src/toplevel-define-type.lisp

```diff
@@ -2,76 +2,137 @@
 
 ;;; Handling of toplevel COALTON:DEFINE-TYPE.
 
-(defun process-toplevel-type-definitions (deftype-forms repr-table env)
-  (declare (type list deftype-forms)
-           (type environment env)
-           (values type-definition-list
-                   environment))
-
-  ;; Parse type definitions into a list of TYPE-DEFINITION objects
-  (let ((parsed-deftypes (parse-type-definitions deftype-forms repr-table env)))
-    (declare (type type-definition-list parsed-deftypes))
+(defparameter *default-eq?-method*
```

Comment:
Is there a reason to manually construct the `make-instance-definition` here instead of generating code for the instance and processing it w/ the hand written instances?

Reply:
A reason and an after-the-fact-justification:
- This way bypasses the `type-definition-explicit-addressable-p` check and error; I couldn't figure out a way to pass a Coalton source s-expression to `process-toplevel-instance-definitions` and friends that wouldn't trigger the error.
- Generating a source form, parsing it and type-checking it would result in a lot of wasted computation for the compiler.

Reply:
I'm not sure about how to best encode the checks. But I would rather keep the compilers internal api surface area small than avoid type checking a one line function. Most of the large structs that are created and eventually fed into a [translation-unit](https://github.com/coalton-lang/coalton/blob/main/src/typechecker/translation-unit.lisp#L3-L10) are only constructed in a single place. This has significantly aided refactoring how data flows through the compiler.

Reply:
Also currently no state is stored outside of the global environment, no part of the compiler is called before it's entirely loaded, and intermediate IRs are never stored between compiler invocations. I would like to maintain all of these properties.

Reply:
These are noble goals, and I'd like to live up to them. The only way I can think of right now is introducing a new form `unchecked-internal-define-instance-yes-i-really-mean-it` that parses like a `define-instance`, but bypasses the check for user-defined instances. I don't love that, but it does preserve the data flow invariants, which seem desirable.

Reply:
Can you process the two lists of instances separately and add a flag to `process-toplevel-instance-definitions` that skips checks for the compiler generated instances?

```
;; in process-coalton-toplevel
(predeclare-toplevel-instances manual-instances env)
(predeclare-toplevel-instances auto-instances env :auto-instances t)

;; type check code

(process-toplevel-instances manual-instances env)
(process-toplevel-instandes auto-instances env :auto-instances t)
```

We've had to do this in the type checker to handle the differences between local and top level bindings. Seee https://github.com/coalton-lang/coalton/blob/main/src/typechecker/derive-type.lisp#L364-L365.

Reply:
Oh, sure, that makes sense.